src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -293,8 +346,66 @@ private TextSpan GetRegionSpanForReference(SourceText sourceText, TextSpan refer
                     sourceText.Lines[lastLineNumber].End);
             }
 
+            private void UpdateCustomColumnsVisibility(ImmutableDictionary<string, ImmutableArray<string>> customData)
+            {
+                // Check if we have any custom reference data to display.
+                if (customData.IsEmpty)
+                {
+                    return;
+                }
+
+                // Get the new column states corresponding to the custom columns to display for custom data.
+                List<ColumnState2> newColumnStatesOpt = null;
+
+                lock (Gate)
+                {
+                    foreach (var customColumnName in customData.Keys)
+                    {
+                        // Get the matching custom column.
+                        var customColumnDefinition = TableControl.ColumnDefinitionManager.GetColumnDefinition(customColumnName) as AbstractFindUsagesCustomColumnDefinition;
+                        if (customColumnDefinition == null)
+                        {
+                            Debug.Fail($"{nameof(SourceReferenceItem.ReferenceInfo)} has a key '{customColumnName}', but there is no exported '{nameof(AbstractFindUsagesCustomColumnDefinition)}' with this name.");
+                            continue;
+                        }
+
+                        // Ensure that we flip the visibility to true for the custom column.
+                        ColumnState2 newColumnStateOpt = null;
+                        if (_customColumnStatesMap.TryGetValue(customColumnDefinition.Name, out var columnState))
+                        {
+                            if (!columnState.IsVisible)
+                            {
+                                newColumnStateOpt = new ColumnState2(columnState.Name, isVisible: true, columnState.Width,
+                                    columnState.SortPriority, columnState.DescendingSort, columnState.GroupingPriority);
+                            }
+                        }
+                        else
+                        {
+                            newColumnStateOpt = customColumnDefinition.DefaultColumnState;
+                        }
+
+                        if (newColumnStateOpt != null)
+                        {
+                            _customColumnStatesMap[customColumnDefinition.Name] = newColumnStateOpt;
+
+                            newColumnStatesOpt = newColumnStatesOpt ?? new List<ColumnState2>();
+                            newColumnStatesOpt.Add(newColumnStateOpt);
+                        }
+                    }
+                }
+
+                // Update the column states if required.
+                if (newColumnStatesOpt != null)
+                {
+                    TableControl.SetColumnStates(newColumnStatesOpt);
```

Comment:
1. is this safe on the BG thread?
2. You might be hearing about this from multiple threads *simultaneously* (find-refs is heavily multi-threaded).  Is this safe?

Reply:
@olegtk?

Reply:
I just verified from looking at Editor layer code that this API [forces switch to UI thread](http://index/?leftProject=Microsoft.VisualStudio.Shell.Framework&leftSymbol=ykzkstkubsjd&rightProject=Microsoft.VisualStudio.Editor.Implementation&file=TableControl%5CTableControl.cs&line=821), so it should be safe to call here. Also note that we will call it only once for each new custom column to add for each FAR query - the lock above this guarantees that `newColumnStatesOpt` is going to be non-null only for the first FAR result that has a non-empty column value.

Reply:
great.  consider doc'ing that to help make it clear this is all intentional and safe.  thanks!